src/ffi/buffer.rs

```diff
@@ -0,0 +1,242 @@
+//! AssemblyScript dynamically sized buffer implementation.
+
+use std::{
+    alloc::{self, Layout, LayoutErr},
+    fmt::{self, Debug, Formatter},
+    marker::PhantomData,
+    mem,
+    ops::Deref,
+    ptr, slice,
+};
+
+/// Internal representation of an AssemblyScript buffer.
+///
+/// `Inner` is declared as a generic struct in order to take advantage of the
+/// partial dynamically sized type (DST) support. For more information see:
+/// <https://doc.rust-lang.org/nomicon/exotic-sizes.html#dynamically-sized-types-dsts>
+#[repr(C)]
+struct Inner<T: ?Sized> {
+    len: usize,
+    buf: T,
+}
+
+/// A borrowed AssemblyScript dynamically sized buffer with elements of type `T`
+/// and aligned to `Alignment`.
+#[repr(transparent)]
+pub struct AscBuf<T, Alignment = T> {
+    _type: PhantomData<T>,
+    inner: Inner<[Alignment; 0]>,
+}
+
+impl<T, A> AscBuf<T, A> {
+    /// Returns the buffer as a Rust slice.
+    pub fn as_slice(&self) -> &[T] {
+        let Inner { len, buf } = &self.inner;
+
+        // SAFETY: `AscBuf` can only be constructed from an `AscBuffer` which
+        // correctly allocates the storage starting at `buf` to have `len`
+        // elements. Additionally we *assume* that all `AscBuf` references from
+        // host calls are valid in the same way.
+        unsafe { slice::from_raw_parts((buf as *const A).cast(), *len) }
+    }
+}
+
+impl<T, A> Debug for AscBuf<T, A>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        Debug::fmt(&self.as_slice(), f)
+    }
+}
+
+/// An owned AssemblyScript dynamically sized buffer with elements of type `T`
+/// and aligned to `Alignment`.
+#[repr(transparent)]
+pub struct AscBuffer<T, Alignment = T> {
+    _type: PhantomData<T>,
+    inner: Inner<[Alignment]>,
+}
+
+impl<T, A> AscBuffer<T, A> {
+    /// Creates a new AssemblyScript buffer from the specified slice.
+    pub fn new(slice: impl AsRef<[T]>) -> Box<Self> {
+        let slice = slice.as_ref();
+
+        // SAFETY: The allocated buffer is guaranteed to be completely
+        // initialized.
+        unsafe {
```

Comment:
It is not entirely clear to me why you need an `unsafe` code block

Reply:
`ptr::copy` and `alloc_buffer` are both `unsafe` functions.